docs/CppInteroperability/Roadmap.md

```diff
@@ -0,0 +1,141 @@
+
+It is a goal to preserve the semantics (including performance semantics) that an API was written with by understanding the API’s semantics and mapping it to a Swift idiom that represents the same thing. 
+
+Swift programmers have used C++ APIs since Swift 1.0 through an Objective-C or C bridging layer. It is important that programmers are able to *incrementally* remove these (potentially huge) bridging layers and start using their C++ APIs directly. To make this transition incremental, C++ interoperability must not change the way Swift imports any existing C or Objective-C APIs, even when C++ interoperability is enabled. 
+
+It is *not* a goal to import every C++ API into Swift. APIs that are imported by Swift should feel Swifty and idiomatic. There must be an easy way to add annotations so that APIs become importable, especially when updating APIs directly is not an option (because the source code is not available). 
+
+## The approach
+
+The approach for importing C++ APIs into Swift is derived from the fact that both our goals, and C++ as a language, necessitate that not all APIs are imported, and to import an API the user must provide semantic information in some cases. Further, in order to realize our goals (especially the goals of safety and clear, well defined mappings), it is impossible to import any arbitrary C++ API. APIs that are imported by Swift must fall into one of our understood categories or be annotated with semantic information that can be used to classify them into a Swift idiom. 
+
+In this section, the roadmap will discuss several of the most common C++ API patterns and how they might be imported by the Swift compiler using the goals outlined above.
+
+### Definitions
+
+**Reference types** have reference semantics and object identity. A reference type is a pointer (or “reference”) to some object which means there is a layer of indirection. When a reference type is copied, the pointer’s value is copied rather than the object’s storage. This means reference types can be used to represent non-copyable types in C++.
+
+**Owned types** “own” some storage which can be copied and destroyed. An owned type must be copyable and destructible. The copy constructor must copy any storage that is owned by the type and the destructor must destroy that storage. Copies and destroys must balance out and these operations must not have side effects. Examples of owned types include `std::vector` and `std::string`. 
+
+**An iterator** represents a point in a range. Iterator types must provide a comparison operator and increment operator. Iterators can be returned by `begin` and `end` methods to form a range.
+
+**Trivial types** can be copied by copying the bits of a value of the trivial type and do not need any special destruction logic. Trivial types own their storage.
```

Comment:
Are trivial types a superset of owned types then? (what relation if any are between these)

Reply:
I believe that this is trying to explain that
```c++
typedef char My32BString[32];
```
would create a trivial type `My32BString`.  The storage is owned by the type and there is no need for a ctor or dtor.

However, if you were to wrap this up into a class e.g.
```c++
struct My32BString {
private:
  char data_[32];
public:
  My32BString(const char *s) {
    memcpy(data_, s, strlen(s));
  }

  My32BString(const My32BString &rhs) {
    memcpy(data_, rhs.data_, sizeof(data_));
  }

  My32BString &operator=(const My32BString &rhs) {
    memcpy(data_, rhs.data_, sizeof(data_);
    return *this;
  }
};
```
it now becomes an owned type as the type constructor must be invoked when creating or copying the value even though it is semantically identical to the trivial type.

An owned type mandates the materialization of a VWT when imported where as a trivial type will not.

Reply:
I see, thanks @compnerd that helped a lot :)

Reply:
Yes, exactly. I think it might actually clarify this to _only_ provide the more detailed definitions below as the added information and type categories should make this distinction clear. WDYT?

Reply:
> An owned type mandates the materialization of a VWT when imported where as a trivial type will not.

This is a good way to put it.

Reply:
Specifically, that means it stops having "swift value type" semantics.